src/DurableTask.Netherite/StorageLayer/Faster/StoreWorker.cs

```diff
@@ -349,7 +371,126 @@ void ScheduleNextIdleCheckpointTime()
+            if (this.pendingStoreCheckpoint != null)
+            {
+                if (this.pendingStoreCheckpoint.IsCompleted == true)
+                {
+                    this.traceHelper.FasterProgress("Checkpointing state machine: store checkpoint no longer pending");
+                    (this.lastCheckpointedCommitLogPosition, this.lastCheckpointedInputQueuePosition)
+                       = await this.pendingStoreCheckpoint; // observe exceptions here
+
+                    // force collection of memory used during checkpointing
+                    GC.Collect();
+
+                    this.traceHelper.FasterProgress("Checkpointing state machine: resetting to initial state");
+                    // we have reached the end of the state machine transitions
+                    this.pendingStoreCheckpoint = null;
+                    this.pendingCheckpointTrigger = CheckpointTrigger.None;
+                    this.ScheduleNextIdleCheckpointTime();
+                    this.partition.Settings.TestHooks?.CheckpointInjector?.SequenceComplete((this.store as FasterKV).Log);
+                }
+                else
+                {
+                    this.traceHelper.FasterProgress("Checkpointing state machine: store checkpoint still pending");
+                }
+            }
+            else if (this.pendingIndexCheckpoint != null)
+            {
+                if (this.pendingIndexCheckpoint.IsCompleted == true)
+                {
+                    this.traceHelper.FasterProgress("Checkpointing state machine: trying to start store checkpoint");
+                    await this.pendingIndexCheckpoint; // observe exceptions here
+
+                    // the store checkpoint is next
+                    this.StartCheckpointOrFailOnTimeout(
+                        checkpointRoutine: () =>
+                        { 
+                            var token = this.store.StartStoreCheckpoint(this.CommitLogPosition, this.InputQueuePosition, this.InputQueueFingerprint, null);
+                            if (token.HasValue)
+                            {
+                                this.traceHelper.FasterProgress("Checkpointing state machine: store checkpoint started");
+                                this.pendingIndexCheckpoint = null;
+                                this.pendingStoreCheckpoint = this.WaitForCheckpointAsync(false, token.Value, true);
+                                this.numberEventsSinceLastCheckpoint = 0;
+                            }
+                            var checkpointStarted = token.HasValue;
+                            return checkpointStarted;
+
+                        },
+                        messageOnError: "Could not start store checkpoint before timeout");
+                }
+                else
+                {
+                    this.traceHelper.FasterProgress("Checkpointing state machine: index checkpoint still pending");
```

Comment:
Same as above; may make more sense to print this info as part of the `LogCheckpointStats`